WalletWasabi/WabiSabi/Client/OwnershipProofValidator.cs

```diff
@@ -0,0 +1,127 @@
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	// OwnershipProofValidator validates the ownership proofs provided by the coordinator.
+	public class OwnershipProofValidator
+	{
+		public OwnershipProofValidator(TransactionStore transactionStore, IBlockProvider blockProvider)
+		{
+			TransactionStore = transactionStore;
+			BlockProvider = blockProvider;
+		}
+
+		public TransactionStore TransactionStore { get; }
+		public IBlockProvider BlockProvider { get; }
+
+		// Verifies the coins and ownership proofs. These proofs are provided by the coordinator who should
+		// have validated them before which means that when dealing with an honest coordinator all these proofs
+		// are valid.
+		//
+		// In case one proof is invalid we have evidence enough to know the coordinator is malicious and we
+		// have to abort the process, notify the user and ban the coordinator.
+		public async ValueTask<int> VerifyOtherAlicesOwnershipProofsAsync(
+			uint256 roundId,
+			IEnumerable<(uint256 BlockId, OutPoint OutPoint, Money Amount, OwnershipProof OwnershipProof)> othersCoins,
+			int minimumNumberOfValidProofs,
+			CancellationToken cancellationToken)
+		{
+			var validProofs = 0;
+			var coinJoinInputCommitmentData = new CoinJoinInputCommitmentData("CoinJoinCoordinatorIdentifier", roundId);
+
+			var mineTxIds = TransactionStore.GetTransactionHashes();
+			var alreadySeenCoins = othersCoins.Where(x => mineTxIds.Contains(x.OutPoint.Hash));
+
+			// In case one Alice is trying to spend an output from a transaction that we already have
+			// seen before then we can verify its script without downloading anything from the network.
+			foreach (var (_, outpoint, amount, ownershipProof) in alreadySeenCoins)
+			{
+				if (validProofs >= minimumNumberOfValidProofs)
+				{
+					break;
+				}
+
+				if (TransactionStore.TryGetTransaction(outpoint.Hash, out var stx))
+				{
+					var transactionCoins = stx.Transaction.Outputs.AsCoins();
+					var foundCoin = transactionCoins.ElementAtOrDefault((int)outpoint.N);
+
+					if (foundCoin is not { } coin || coin.Amount != amount)
+					{
+						throw new InvalidOperationException("The coordinator lies (coin doesn't exists or is different than the provided by the coordinator).");
+					}
+					VerifyCoin(foundCoin.ScriptPubKey, coinJoinInputCommitmentData, ownershipProof);
+					validProofs++;
+				}
+			}
+
+			// In case we cannot validate enough proofs using only our already seen transactions, we
+			// would need to start downloading the blocks containing the coins that need to be
+			// validated.
+			var coinsGroupedByBlocks = othersCoins.GroupBy(x => x.BlockId).ToArray();
+
+			// We sort the groups (blocks) randomly.
+			var random = new Random();
+			var randomNumbers = othersCoins.Select(r => random.Next()).ToArray();
+			coinsGroupedByBlocks = coinsGroupedByBlocks.Zip(randomNumbers, (group, rnd) => new { Groupd = group, Order = rnd })
+				.OrderBy(x => x.Order)
+				.Select(x => x.Groupd)
+				.ToArray();
```

Comment:
isn't there a `Shuffle()` method?

Reply:
How about this? 

`coinsGroupedByBlocks = coinsGroupedByBlocks.OrderBy(a => Guid.NewGuid()).ToArray();`

Reply:
There's an extension method defined on `List`, it could be improved (to take an optional `IRandom` and to also work on arrays or `IEnumerable`s perhaps) but anyway I think this is cleanest for now (see also above suggestion)

```suggestion
			coinsGroupedByBlocks.Shuffle()
```

Reply:
Replaced.